### PR TITLE
use at::Device in JIT

### PR DESCRIPTION
**Pull request by ArmenAg**
This pull request will be used to monitor the progress on using `at::Device` in JIT instead of `int`.

@zdevito If you don't mind, any question I have I'll ask on this thread.

### CODE REVIEW

**Comment by ArmenAg** [comment]
The next thing I will attempt to change is the use of `int device` in `compiler.h` and `executor.h` in `torch/csrs/jit/fuser`.

File: aten/src/ATen/core/jit_type.h [hunk 1]
```diff
@@ -1,69 +1,72 @@
...
 };
 
-#define DEFINE_IS_SUBCLASS(_kind) \
+#define DEFINE_IS_SUBCLASS(_kind)                       \
   bool isSubclass(const TypeKind kind) const override { \
-    return kind == TypeKind::_kind; \
+    return kind == TypeKind::_kind;                     \
   }
 
+inline static at::Device ConvertIntToDevice(int device) {
```
Comment (ezyang) [inline aten/src/ATen/core/jit_type.h]: This is very CUDA-specific; put CUDA in its name? :)

**Comment by ezyang** [comment]
@ArmenAg, as a general note, try to avoid superfluous whitespace changes when making a refactoring change like this; every extra line you touch is an extra one you might be merge conflicting with someone else. We're happy to take reformatting, but it's best if it lives in a PR of its own.

File: aten/src/ATen/core/jit_type.h [hunk 2]
```diff
@@ -339,52 +362,106 @@ struct CAFFE2_API TensorType : public DynamicType {
...
+      bool requires_grad = true,
+      TypeKind kind = TypeKind::TensorType)
+      : DynamicType(kind),
+        scalar_type_(scalar_type),
+        requires_grad_(at::isFloatingType(scalar_type) && requires_grad),
+        device_(device),
+        dim_(dim) {}
+  TensorType(
+      at::ScalarType scalar_type,
+      int device,
```
Comment (ezyang) [inline aten/src/ATen/core/jit_type.h]: Shouldn't this take an `at::Device`?

File: aten/src/ATen/core/jit_type.h [hunk 3]
```diff
@@ -437,25 +511,70 @@ struct CAFFE2_API CompleteTensorType : public TensorType {
...
+            device,
+            sizes.size(),
+            requires_grad,
+            TypeKind::CompleteTensorType),
+        sizes_(sizes.vec()),
+        strides_(strides.vec()) {}
+
+  CompleteTensorType(
+      at::ScalarType scalar_type,
+      int device,
```
Comment (ezyang) [inline aten/src/ATen/core/jit_type.h]: Kill this constructor?

File: aten/src/ATen/core/jit_type.h [hunk 4]
```diff
@@ -437,25 +511,70 @@ struct CAFFE2_API CompleteTensorType : public TensorType {
...
+      bool requires_grad = true)
+      : CompleteTensorType(
+            scalar_type,
+            ConvertIntToDevice(device),
+            sizes,
+            CompleteTensorType::contiguousStridesOf(sizes),
+            requires_grad) {}
+  CompleteTensorType(
+      at::ScalarType scalar_type,
+      int device,
```
Comment (ezyang) [inline aten/src/ATen/core/jit_type.h]: This one too

File: torch/csrc/jit/python_arg_flatten.h [hunk 5]
```diff
@@ -61,26 +63,31 @@ struct IODescriptor {
...
+static inline std::ostream& operator<<(
+    std::ostream& out,
+    const IODescriptor::VariableMetadata& meta) {
+  auto& t = at::getNonVariableType(
+      meta.device.is_cpu() ? at::Backend::CPU : at::Backend::CUDA, meta.type);
   out << t << "(requires_grad=" << meta.requires_grad;
-  if (meta.device > 0) {
-    out << ", device=" << meta.device;
+  if (meta.device.index() > 0) {
+    out << ", device=" << meta.device.index();
```
Comment (ezyang) [inline torch/csrc/jit/python_arg_flatten.h]: This looks CUDA-specific

File: torch/csrc/jit/ir.cpp [hunk 6]
```diff
@@ -198,7 +198,7 @@ void Graph::dumpPretty() {
 
 static void checkSameDevice(const Node* node) {
   bool has_device = false;
-  int device;
+  at::Device device = at::kCPU;
```
Comment (ezyang) [inline torch/csrc/jit/ir.cpp]: Use optional<Device> here instead of the bool and int

**Comment by ArmenAg** [comment]
@ezyang Thank you for your comments. I'll fix your comments today or over the weekend at the latest. Sorry about the whitespace issue, I didn't realize it until it was to late. Should I submit a PR containing reformatting for all the jit code? Would this be useful?

**Comment by zdevito** [comment]
@ArmenAg re:whitespace: we really can't easily merge formatting changes, they will mess up a lot of people's inflight work. It also makes the diff extremely difficult to read. Is it possible for you to resend the diff with only the non-formatting changes?

**Comment by ArmenAg** [comment]
@zdevito Sure I can do that. I'll probably send in a different PR, that will be a bit easier for me. What should we do in terms of formatting in the future? There seems to be a lot of variation on how people format their code where auto-formatting can help. There's especially a bit of inconsistency in terms of white spacing.

**Comment by ArmenAg** [comment]
Moving this to another pull request.

**Pull request closed without merging**